src/tests/stats/test_median.fypp

```diff
@@ -0,0 +1,234 @@
+    call check( abs(median(d2) - 1._dp) < dptol, '${k1}$ median(d2), odd')
+
+    d3 = reshape(d1, [1, 3, 5], pad = [0_${k1}$])
+    call check(mod(size(d3), 2) == 1, 'test_median_${k1}$: should be an odd number')
+    call check( abs(median(d3) - 1._dp) < dptol, '${k1}$ median(d3), odd')
+
+    call check( abs(median(d1, 1) - 2._dp) < dptol, '${k1}$ median(d1, 1)')
+    call check( sum(abs(median(d2, 1) - [2._dp, -4._dp, 7._dp, 1._dp, 0._dp])) < dptol, &
+                 '${k1}$ median(d2, 1), odd')
+    call check( sum(abs(median(d2, 2) - [7._dp, 1._dp, 0._dp])) < dptol,                &
+                 '${k1}$ median(d2, 2), odd')
+
+end subroutine
+#:endfor
+
+#:for k1,t1 in REAL_KINDS_TYPES
+subroutine test_median_${k1}$()
+    ${t1}$, allocatable :: d0(:), d1(:), d2(:,:), d3(:,:,:)
+    ${t1}$, allocatable :: tmp1(:)
+
+    allocate(d0(0))
+    !check just to be sure that the setup of d0 is correct
+    call check(size(d0) == 0, 'test_median_${k1}$: should be of size 0')
+
+    d1 = [10., 2., -3., -4., 6., -6., 7., -8., 9., 0., 1., 20.]
+    call check(mod(size(d1), 2) == 0, 'test_median_${k1}$: should be an even number')
+
+    d2 = reshape(d1, [3, 4])
+    call check(mod(size(d2), 2) == 0, 'test_median_${k1}$: should be an even number')
+
+    d3 = reshape(d1, [2, 3, 2])
+    call check(mod(size(d3), 2) == 0, 'test_median_${k1}$: should be an even number')
+
+    !median_all
+    call check( ieee_is_nan(median(d0)), '${k1}$ median(d0)' )
+    call check( ieee_is_nan(median(d0, .false.)), '${k1}$ median(d0, .false.)' )
+    call check( ieee_is_nan(median(d1, .false.)), '${k1}$ median(d1, .false.)' )
+    call check( ieee_is_nan(median(d2, .false.)), '${k1}$ median(d2, .false.)' )
+    call check( ieee_is_nan(median(d3, .false.)), '${k1}$ median(d3, .false.)' )
+
+    call check( abs(median(d1) - 1.5_${k1}$) < ${k1}$tol, '${k1}$ median(d1), even')
+    call check( abs(median(d2) - 1.5_${k1}$) < ${k1}$tol, '${k1}$ median(d2), even')
+    call check( abs(median(d3) - 1.5_${k1}$) < ${k1}$tol, '${k1}$ median(d3), even')
+
+    !median
+    call check( ieee_is_nan(median(d0, 1)), '${k1}$ median(d0, 1)' )
+    call check( ieee_is_nan(median(d0, 1, .false.)), '${k1}$ median(d0, 1, .false.)' )
+    call check( ieee_is_nan(median(d1, 1, .false.)), '${k1}$ median(d1, 1, .false.)' )
+    call check( any(ieee_is_nan(median(d2, 1, .false.))), '${k1}$ median(d2, 1, .false.)' )
+    call check( any(ieee_is_nan(median(d2, 2, .false.))), '${k1}$ median(d2, 2, .false.)' )
+    call check( any(ieee_is_nan(median(d3, 1, .false.))), '${k1}$ median(d3, 1, .false.)' )
+    call check( any(ieee_is_nan(median(d3, 2, .false.))), '${k1}$ median(d3, 2, .false.)' )
+    call check( any(ieee_is_nan(median(d3, 3, .false.))), '${k1}$ median(d3, 3, .false.)' )
+
+    call check( abs(median(d1, 1) - 1.5_${k1}$) < ${k1}$tol, '${k1}$ median(d1, 1), even')
+    call check( sum(abs(median(d2, 1) - [2._${k1}$, -4._${k1}$, 7._${k1}$, 1._${k1}$])) < ${k1}$tol, &
```

Comment:
Would using the array kind specifier help reduce the preprocessor noise?

```
[real(${k1}$) :: 2.0, -4.0, 7.0, 1.0]
```

Reply:
With `gfortran`, this implies a conversion from `real(4)` to `real(8)`, with a warning triggered by `-Wconversion-extra`.
For this reason, I am inclined to keep as it is now.